### PR TITLE
✨ [Bocato] Add HomeViewModel Tests

### PR DESCRIPTION
**Pull request by mr-araujo**
### Description

Add HomeViewModel Tests 
 
### Checklist:
- [x] No duplicate code
- [x] No comment code
- [x] No WIP code

### CODE REVIEW

File: solutions/devsprint-eduardo-bocato-2/FinanceAppTests/HomeViewModelTests.swift [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    override func tearDown() {
+        super.tearDown()
+        sut = nil
+    }
+
+    func test_fetchData_withData_shouldCallWithSuccess() throws {
+        sut.fetchData()
+        XCTAssertTrue(delegate.displayedHomeData)
```
Comment (bocato) [inline solutions/devsprint-eduardo-bocato-2/FinanceAppTests/HomeViewModelTests.swift:30]: **[CHANGE REQUEST]**
Please apply G/W/T.

File: solutions/devsprint-eduardo-bocato-2/FinanceAppTests/HomeViewModelTests.swift [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    func test_fetchData_withoutData_shouldCallWithError() throws {
+        service.errorAPI = .noData
+        sut.fetchData()
+        XCTAssertFalse(delegate.displayedError)
```
Comment (bocato) [inline solutions/devsprint-eduardo-bocato-2/FinanceAppTests/HomeViewModelTests.swift:36]: **[CHANGE REQUEST]**
Please apply G/W/T.

File: solutions/devsprint-eduardo-bocato-2/FinanceAppTests/Mock/FinanceServiceMock.swift [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+            case .parseError:
+                completion(nil)
+            case .noData:
+                completion(nil)
+        default:
+            let homeData = HomeData(balance: 100, savings: 500, spending: 100, activity: [])
+            completion(homeData)
+        }
+    }
+}
```
Comment (bocato) [inline solutions/devsprint-eduardo-bocato-2/FinanceAppTests/Mock/FinanceServiceMock.swift:30]: **[CHANGE REQUEST]**
Please use a stub and configure the scenarios in your test functions.
```suggestion
final class  FinanceServiceStub: FinanceServiceProtocol {
    var homeDataToBeReturned: HomeData? = nil
    func fetchHomeData(_ completion: @escaping (HomeData?) -> Void) {
        completion(homeDataToBeReturned)
    }
}
```

File: solutions/devsprint-eduardo-bocato-2/FinanceAppTests/Mock/HomeViewModelDelegateMock.swift [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+class HomeViewModelDelegateMock: HomeViewModelDelegate {
+
+    var displayedHomeData = false
+    var displayedError = false
+
+    func didFetchHomeData(_ data: HomeData) {
+        displayedHomeData = true
+    }
+}
```
Comment (bocato) [inline solutions/devsprint-eduardo-bocato-2/FinanceAppTests/Mock/HomeViewModelDelegateMock.swift:19]: **[CHANGE REQUEST]**
Please use a spy and configure the scenarios in your test functions.
```suggestion
final class HomeViewModelDelegateSpy: HomeViewModelDelegate {
    private(set) var didFetchHomeDataCalled = false
    private(set) var homeDataPassed: HomeData?

    func didFetchHomeData(_ data: HomeData) {
        didFetchHomeDataCalled = true
        homeDataPassed = data
    }
}
```

**Review by bocato — CHANGES_REQUESTED** [review]

File: solutions/devsprint-eduardo-bocato-2/FinanceApp/Scenes/Home/HomeViewModel.swift [hunk 5]
```diff
@@ -7,12 +7,18 @@ protocol HomeViewModelDelegate: AnyObject {
...
+    init(with service: FinanceServiceProtocol) {
+        self.service = service
+    }
+
     func fetchData() {
-        FinanceService.fetchHomeData { homeData in
+        service.fetchHomeData { homeData in
             guard let homeData = homeData else { return }
-            DispatchQueue.main.async {
+//            DispatchQueue.main.async {
```
[the remark this answers is not in the document]
Comment (mr-araujo) [reply solutions/devsprint-eduardo-bocato-2/FinanceApp/Scenes/Home/HomeViewModel.swift]: Nice, fixed

File: solutions/devsprint-eduardo-bocato-2/FinanceAppTests/HomeViewModelTests.swift (hunk 1 above)
Comment (mr-araujo) [reply solutions/devsprint-eduardo-bocato-2/FinanceAppTests/HomeViewModelTests.swift]: Done

File: solutions/devsprint-eduardo-bocato-2/FinanceAppTests/HomeViewModelTests.swift [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        sut.delegate = delegate
+        super.setUp()
+    }
+
+    override func tearDown() {
+        super.tearDown()
+        sut = nil
+    }
+
+    func test_fetchData_withData_shouldCallWithSuccess() throws {
```
[the remark this answers is not in the document]
Comment (mr-araujo) [reply solutions/devsprint-eduardo-bocato-2/FinanceAppTests/HomeViewModelTests.swift]: Good suggestion

File: solutions/devsprint-eduardo-bocato-2/FinanceAppTests/HomeViewModelTests.swift (hunk 2 above)
Comment (mr-araujo) [reply solutions/devsprint-eduardo-bocato-2/FinanceAppTests/HomeViewModelTests.swift]: Done

File: solutions/devsprint-eduardo-bocato-2/FinanceAppTests/Mock/FinanceServiceMock.swift (hunk 3 above)
Comment (mr-araujo) [reply solutions/devsprint-eduardo-bocato-2/FinanceAppTests/Mock/FinanceServiceMock.swift]: Done

File: solutions/devsprint-eduardo-bocato-2/FinanceAppTests/Mock/HomeViewModelDelegateMock.swift (hunk 4 above)
Comment (mr-araujo) [reply solutions/devsprint-eduardo-bocato-2/FinanceAppTests/Mock/HomeViewModelDelegateMock.swift]: Done

**Review by bocato — APPROVED** [review]

**Pull request merged**